network/kademlialoadbalancer.go

```diff
@@ -0,0 +1,334 @@
+		peers := klb.peerBinToPeerList(peerBin)
+		return consumeBin(LBBin{LBPeers: peers, ProximityOrder: peerBin.ProximityOrder})
+	})
+}
+
+// EachBinDesc returns all bins in descending order from the perspective of base address.
+// All peers in that bin will be provided to the LBBinConsumer sorted by least used first.
+func (klb *KademliaLoadBalancer) EachBinDesc(base []byte, consumeBin LBBinConsumer) {
+	klb.kademlia.EachBinDesc(base, 0, func(peerBin *PeerBin) bool {
+		peers := klb.peerBinToPeerList(peerBin)
+		return consumeBin(LBBin{LBPeers: peers, ProximityOrder: peerBin.ProximityOrder})
+	})
+}
+
+func (klb *KademliaLoadBalancer) peerBinToPeerList(bin *PeerBin) []LBPeer {
+	resources := make([]Resource, bin.Size)
+	var i int
+	bin.PeerIterator(func(entry *entry) bool {
+		resources[i] = entry.conn
+		i++
+		return true
+	})
+	return klb.resourcesToLbPeers(resources)
+}
+
+func (klb *KademliaLoadBalancer) resourcesToLbPeers(resources []Resource) []LBPeer {
+	sorted := klb.resourceUseStats.sortResources(resources)
+	peers := klb.toLBPeers(sorted)
+	return peers
+}
+
+func (klb *KademliaLoadBalancer) listenNewPeers() {
+	for {
+		select {
+		case <-klb.quitC:
+			return
+		case msg, ok := <-klb.onPeerSub.ReceiveChannel():
+			if !ok {
+				return
+			}
+			signal, ok := msg.(newPeerSignal)
+			if !ok {
+				return
+			}
+			klb.addedPeer(signal.peer, signal.po)
+		}
+	}
+}
+
+func (klb *KademliaLoadBalancer) listenOffPeers() {
+	for {
+		select {
+		case <-klb.quitC:
+			return
+		case msg := <-klb.offPeerSub.ReceiveChannel():
+			peer, ok := msg.(*Peer)
+			if peer != nil && ok {
+				klb.removedPeer(peer)
+			}
+		}
+	}
+}
+
+// addedPeer is called back when a new peer is added to the kademlia. Its uses will be initialized
+// to the use count of the least used peer in its bin. The po of the new peer is passed to avoid having
+// to calculate it again.
+func (klb *KademliaLoadBalancer) addedPeer(peer *Peer, po int) {
+	initCount := klb.initCountFunc(peer, 0)
+	log.Debug("Adding peer", "key", peer.Label(), "initCount", initCount)
+	klb.resourceUseStats.initKey(peer.Key(), initCount)
+}
+
+// leastUsedCountInBin returns the use count for the least used peer in this bin excluding the excludePeer.
+func (klb *KademliaLoadBalancer) leastUsedCountInBin(excludePeer *Peer, po int) int {
+	addr := klb.kademlia.BaseAddr()
+	peersInSamePo := klb.getPeersForPo(addr, po)
+	idx := 0
+	leastUsedCount := 0
+	for idx < len(peersInSamePo) {
+		leastUsed := peersInSamePo[idx]
+		if leastUsed.Peer.Key() != excludePeer.Key() {
+			leastUsedCount = klb.resourceUseStats.getUses(leastUsed.Peer)
+			log.Debug("Least used peer is", "peer", leastUsed.Peer.Label(), "leastUsedCount", leastUsedCount)
+			break
+		}
+		idx++
+	}
+	return leastUsedCount
+}
+
+// nearestNeighbourCount returns the use count for the closest peer count.
+func (klb *KademliaLoadBalancer) nearestNeighbourCount(newPeer *Peer, _ int) int {
```

Comment:
I like the suggested name, I will change it.
The second parameter is ignored because this function is a initCountFunc whose type is `initCountFunc func(peer *Peer, po int) int`. In this case is not used but for `leastUsedCountInBin` it is.

New peer is a recently added peer to kademlia. This function initializes the `useCount` of this new peer.